# Software Coaching for Python
# Week 3: Pandas Dataframes - Homework

Instructor: Kang-Pyo Lee 

In [2]:
from google.colab import drive

drive.mount('/content/gdrive')
path = "/content/gdrive/Shareddrives/SW_Coaching_for_Python"
my_folder = "ESK"     # *** REPLACE WITH YOUR FOLDER NAME ***
outcome_folder = f"{path}/{my_folder}/outcome"
classdata_folder = f"{path}/classdata"

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


## Element Selection from a Dataframe

In [3]:
import pandas as pd
pd.set_option('display.max_colwidth', 150)

df = pd.read_csv(f"{classdata_folder}/timeline_NASA.csv", sep="\t")
df

status_id  ...              retweet_created_at
0     1354552906175934477  ...                             NaN
1     1354549784061865986  ...                             NaN
2     1354546326139265030  ...                             NaN
3     1354542309988327425  ...                             NaN
4     1354540246436872195  ...  Wed Jan 27 20:18:56 +0000 2021
...                   ...  ...                             ...
3195  1265370733930307585  ...  Tue May 26 19:36:00 +0000 2020
3196  1265365848098816002  ...                             NaN
3197  1265327291468562432  ...                             NaN
3198  1265327258274840576  ...                             NaN
3199  1265317908068085761  ...  Tue May 26 16:09:54 +0000 2020

[3200 rows x 7 columns]

1\. Get the list of column labels of `df`. 

In [4]:
# Your answer here
df.columns

Index(['status_id', 'created_at', 'text', 'retweet_count', 'favorite_count',
       'is_retweet', 'retweet_created_at'],
      dtype='object')

2\. Get the list of row index positions of `df`. 

In [6]:
# Your answer here
list(df.index)

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


3\. Get the shape of `df`. 

In [7]:
# Your answer here
df.shape

(3200, 7)

4\. Get the number of rows in `df`. 

In [8]:
# Your answer here
df.shape[0]
# len(df)

3200

5\. Select all values from the `created_at` column. 

In [9]:
# Your answer here
df.created_at

0       Wed Jan 27 22:12:33 +0000 2021
1       Wed Jan 27 22:00:09 +0000 2021
2       Wed Jan 27 21:46:25 +0000 2021
3       Wed Jan 27 21:30:27 +0000 2021
4       Wed Jan 27 21:22:15 +0000 2021
                     ...              
3195    Tue May 26 19:54:27 +0000 2020
3196    Tue May 26 19:35:02 +0000 2020
3197    Tue May 26 17:01:49 +0000 2020
3198    Tue May 26 17:01:41 +0000 2020
3199    Tue May 26 16:24:32 +0000 2020
Name: created_at, Length: 3200, dtype: object

6\. Select the first 3 rows from `df`. 

In [10]:
# Your answer here
df.head(3)

status_id  ... retweet_created_at
0  1354552906175934477  ...                NaN
1  1354549784061865986  ...                NaN
2  1354546326139265030  ...                NaN

[3 rows x 7 columns]

7\. Select the last 3 rows from `df`.

In [11]:
# Your answer here
df.tail(3)

status_id  ...              retweet_created_at
3197  1265327291468562432  ...                             NaN
3198  1265327258274840576  ...                             NaN
3199  1265317908068085761  ...  Tue May 26 16:09:54 +0000 2020

[3 rows x 7 columns]

8. Select the element at row index position 999 and in the `text` column.

In [12]:
# Your answer here
df.text[999]

'RT @NASAMoon: Happy Halloween🎃! A full Moon is always a treat, but this one is a trickster👻 As the second full Moon in a month, we call i…'

9\. Select the element in the fifth row and the third column.

In [13]:
# Your answer here
df.iloc[4][2]

"RT @NASAMars: Don't forget to join us for today's briefing at 4:30pm ET (21:30 UTC): https://t.co/fGItMYWnFR Tag your questions #Countdown…"

10\. Select all values from the last column, without specifying the column label.

In [14]:
# Your answer here
df.iloc[:, -1]

0                                  NaN
1                                  NaN
2                                  NaN
3                                  NaN
4       Wed Jan 27 20:18:56 +0000 2021
                     ...              
3195    Tue May 26 19:36:00 +0000 2020
3196                               NaN
3197                               NaN
3198                               NaN
3199    Tue May 26 16:09:54 +0000 2020
Name: retweet_created_at, Length: 3200, dtype: object

11\. Select all rows in the last two columns, without specifying the column labels.

In [15]:
# Your answer here
df.iloc[:,-2:]

is_retweet              retweet_created_at
0              0                             NaN
1              0                             NaN
2              0                             NaN
3              0                             NaN
4              1  Wed Jan 27 20:18:56 +0000 2021
...          ...                             ...
3195           1  Tue May 26 19:36:00 +0000 2020
3196           0                             NaN
3197           0                             NaN
3198           0                             NaN
3199           1  Tue May 26 16:09:54 +0000 2020

[3200 rows x 2 columns]

12\. Select the first five rows in the last two columns, without specifying the column labels.

In [16]:
# Your answer here
df.iloc[:5,-2:]

is_retweet              retweet_created_at
0           0                             NaN
1           0                             NaN
2           0                             NaN
3           0                             NaN
4           1  Wed Jan 27 20:18:56 +0000 2021

13\. Select all rows from `df` with the `favorite_count` column value being strictly larger than 5000.

In [17]:
# Your answer here
df[df.favorite_count > 5000]

status_id  ... retweet_created_at
32    1354376310638977025  ...                NaN
34    1354246577523740674  ...                NaN
43    1354102852323520513  ...                NaN
52    1353431128766603264  ...                NaN
56    1353081323578150912  ...                NaN
...                   ...  ...                ...
3168  1265655788858748929  ...                NaN
3171  1265643951513624579  ...                NaN
3174  1265577232728285190  ...                NaN
3181  1265478732019650562  ...                NaN
3194  1265386899197308928  ...                NaN

[512 rows x 7 columns]

14\. Select all rows from `df` with the `favorite_count` column value being between 5000 (inclusive) and 10000 (exclusive).

In [19]:
# Your answer here
df[(df.favorite_count >= 5000) & (df.favorite_count < 10000)]

status_id  ... retweet_created_at
32    1354376310638977025  ...                NaN
34    1354246577523740674  ...                NaN
43    1354102852323520513  ...                NaN
52    1353431128766603264  ...                NaN
59    1352754920852697088  ...                NaN
...                   ...  ...                ...
3140  1265692470869925888  ...                NaN
3155  1265685570564603904  ...                NaN
3156  1265685157257842690  ...                NaN
3171  1265643951513624579  ...                NaN
3194  1265386899197308928  ...                NaN

[333 rows x 7 columns]

15\. Create a random sample of 100 rows from `df` without duplicates. 

In [20]:
# Your answer here
df.sample(n = 100, replace=False,random_state=0)

status_id  ...              retweet_created_at
536   1333192674883592193  ...  Sun Nov 29 22:36:06 +0000 2020
1716  1298701153899810818  ...                             NaN
499   1334900692603965441  ...  Fri Dec 04 15:58:26 +0000 2020
567   1331676497069842433  ...  Wed Nov 25 18:25:30 +0000 2020
2567  1278343229876903939  ...                             NaN
...                   ...  ...                             ...
1763  1296478350924488707  ...  Thu Aug 20 15:09:02 +0000 2020
3133  1265696086083227648  ...                             NaN
2737  1273711084193619970  ...                             NaN
2549  1278450641409642498  ...                             NaN
2361  1284997702393233410  ...  Sun Jul 19 23:41:02 +0000 2020

[100 rows x 7 columns]

## Dataframe Iteration & File Writing

Let's continue to use `df`.

1\. Select the first ten values from the `text` column of `df` and print the first 30 characters of each value concatenated by the string '...'. Rows in the printed output are separated by a new line. For example, the first three rows in the printed output look as follows.

It once held a lake. And what ...<br>
I asked the team [...] to come...<br>
Perseverance is the very first...<br>

In [21]:
# Your answer here
for text in df[:10].text:
    print(text[:30] + "...")

It once held a lake. And what ...
I asked the team [...] to come...
Perseverance is the very first...
🔴 LIVE: Tune in for a preview ...
RT @NASAMars: Don't forget to ...
Just 22 more days until we #Co...
RT @NASAExoplanets: 📣Discovery...
RT @Space_Station: .@Astro_Ill...
That's a wrap! @Astro_illini a...
Success! With a bit of strengt...


2\. Select the first ten rows from `df` and print the values of each row in the following format: 'created at: CREATED_AT, text: TEXT, retweet_count: RETWEET_COUNT', where CREATED_AT, TEXT, and RETWEET_COUNT are the values from the three columns `created_at`, `text`, and `retweet_count`, respectively. Rows are separated by two new lines. For example, the first three rows in the printed output look as follows. 

created_at: Wed Jan 27 22:12:33 +0000 2021, text: It once held a lake. And what we're going to do, as soon as we get there, is to explore the rocks that were deposi… https://t.co/UW9wngvx4C, retweet_count: 72<br>

created_at: Wed Jan 27 22:00:09 +0000 2021, text: I asked the team [...] to come up with something that would symbolize, to mark this challenge and thank in particu… https://t.co/QH235r9zjq, retweet_count: 50<br>

created_at: Wed Jan 27 21:46:25 +0000 2021, text: Perseverance is the very first rover designed to seek signs of past microbial life, by collecting and caching rock… https://t.co/nn72OHRVv6, retweet_count: 135<br>

In [42]:
# Your answer here
df2 = df[:10]
for idx, rows in df2.iterrows():
  print(f"created_at: {rows.created_at} text: {rows.text} retweet_count: {rows.retweet_count}")

created_at: Wed Jan 27 22:12:33 +0000 2021 text: It once held a lake. And what we're going to do, as soon as we get there, is to explore the rocks that were deposi… https://t.co/UW9wngvx4C retweet_count: 72
created_at: Wed Jan 27 22:00:09 +0000 2021 text: I asked the team [...] to come up with something that would symbolize, to mark this challenge and thank in particu… https://t.co/QH235r9zjq retweet_count: 50
created_at: Wed Jan 27 21:46:25 +0000 2021 text: Perseverance is the very first rover designed to seek signs of past microbial life, by collecting and caching rock… https://t.co/nn72OHRVv6 retweet_count: 135
created_at: Wed Jan 27 21:30:27 +0000 2021 text: 🔴 LIVE: Tune in for a preview of the next Mars landing! On Feb. 18, our @NASAPersevere rover will arrive at the Red… https://t.co/rHRhIrzXWR retweet_count: 374
created_at: Wed Jan 27 21:22:15 +0000 2021 text: RT @NASAMars: Don't forget to join us for today's briefing at 4:30pm ET (21:30 UTC): https://t.co/fGItMYWnFR Tag your qu

3\. Select all rows from `df` with the `is_retweet` column value being 0 and save the resulting dataframe in a CSV file named `hw3_output.csv` in the `outcome` folder such that the column separator is a tab and it contains the index of the dataframe. 

In [43]:
# Your answer here
df3 = df[df.is_retweet == 0]
df3.to_csv(f"{outcome_folder}/hw3_output.csv", sep = '\t', index = False)

## Adding Columns to a Dataframe

Let's continue to use `df`.

1\. Add a new column `text_shortened` to `df` such that each value in the column is the first 30 characters of the `text` column value concatenated by the string '...'.  

In [44]:
# Your answer here
df["text_shortened"] = df.text.apply(lambda x: x[:30] + "...")

In [45]:
# Check your answer here.
df[["text", "text_shortened"]]

text                     text_shortened
0      It once held a lake. And what we're going to do, as soon as we get there, is to explore the rocks that were deposi… https://t.co/UW9wngvx4C  It once held a lake. And what ...
1      I asked the team [...] to come up with something that would symbolize, to mark this challenge and thank in particu… https://t.co/QH235r9zjq  I asked the team [...] to come...
2      Perseverance is the very first rover designed to seek signs of past microbial life, by collecting and caching rock… https://t.co/nn72OHRVv6  Perseverance is the very first...
3     🔴 LIVE: Tune in for a preview of the next Mars landing! On Feb. 18, our @NASAPersevere rover will arrive at the Red… https://t.co/rHRhIrzXWR  🔴 LIVE: Tune in for a preview ...
4      RT @NASAMars: Don't forget to join us for today's briefing at 4:30pm ET (21:30 UTC): https://t.co/fGItMYWnFR Tag your questions #Countdown…  RT @NASAMars: Don't forget to ...
...                                                                                                                                            ...                                ...
3195    RT @NASAKennedy: Looking for live views of the launch pad? We got you covered. 🚀👀Check out the rocket that will carry @AstroBehnken and @…  RT @NASAKennedy: Looking for l...
3196  For the first time since 2011, we’re sending American astronauts back to space, on an American rocket, from America… https://t.co/FsC1feO84c  For the first time since 2011,...
3197    LIVE: Want to go behind-the-scenes of #LaunchAmerica?Join our experts on @reddit to ask questions about the missi… https://t.co/p7iqJY4jyX  LIVE: Want to go behind-the-sc...
3198                                                               @brandonleblanc @SpaceX @AmyShiraTeitel See you online! https://t.co/FaQ7K75sBP  @brandonleblanc @SpaceX @AmySh...
3199  RT @NASAEarth: Ready to #LaunchAmerica? Tomorrow, two astronauts 👨🏻‍🚀👨🏻‍🚀 are launching 🚀 to the @Space_Station from @NASAKennedy’s histori…  RT @NASAEarth: Ready to #Launc...

[3200 rows x 2 columns]

2\. Add a new column `first_last_token` to `df` such that each column value is a tuple with two elements: the first element being the first token of the `text` column value and the second one being the last token (use the <b>split</b> method). 

In [46]:
# Your answer here
df["first_last_token"] = df.text.apply(lambda x: (x.split(' ')[0], x.split(' ')[-1]))

In [47]:
# Check your answer here.
df[["text", "first_last_token"]]

text                            first_last_token
0      It once held a lake. And what we're going to do, as soon as we get there, is to explore the rocks that were deposi… https://t.co/UW9wngvx4C               (It, https://t.co/UW9wngvx4C)
1      I asked the team [...] to come up with something that would symbolize, to mark this challenge and thank in particu… https://t.co/QH235r9zjq                (I, https://t.co/QH235r9zjq)
2      Perseverance is the very first rover designed to seek signs of past microbial life, by collecting and caching rock… https://t.co/nn72OHRVv6     (Perseverance, https://t.co/nn72OHRVv6)
3     🔴 LIVE: Tune in for a preview of the next Mars landing! On Feb. 18, our @NASAPersevere rover will arrive at the Red… https://t.co/rHRhIrzXWR                (🔴, https://t.co/rHRhIrzXWR)
4      RT @NASAMars: Don't forget to join us for today's briefing at 4:30pm ET (21:30 UTC): https://t.co/fGItMYWnFR Tag your questions #Countdown…                           (RT, #Countdown…)
...                                                                                                                                            ...                                         ...
3195    RT @NASAKennedy: Looking for live views of the launch pad? We got you covered. 🚀👀Check out the rocket that will carry @AstroBehnken and @…                                    (RT, @…)
3196  For the first time since 2011, we’re sending American astronauts back to space, on an American rocket, from America… https://t.co/FsC1feO84c              (For, https://t.co/FsC1feO84c)
3197    LIVE: Want to go behind-the-scenes of #LaunchAmerica?Join our experts on @reddit to ask questions about the missi… https://t.co/p7iqJY4jyX            (LIVE:, https://t.co/p7iqJY4jyX)
3198                                                               @brandonleblanc @SpaceX @AmyShiraTeitel See you online! https://t.co/FaQ7K75sBP  (@brandonleblanc, https://t.co/FaQ7K75sBP)
3199  RT @NASAEarth: Ready to #LaunchAmerica? Tomorrow, two astronauts 👨🏻‍🚀👨🏻‍🚀 are launching 🚀 to the @Space_Station from @NASAKennedy’s histori…                              (RT, histori…)

[3200 rows x 2 columns]

3\. Add a new column `metrics` to `df` such that each column value is a tuple with three values from the three columns `retweet_count`, `favorite_count`, and `is_retweet`, respectively, in order. 

In [49]:
# Your answer here
df["metrics"] = df.apply(lambda x: (x.retweet_count, x.favorite_count, x.is_retweet), axis = 1)

In [50]:
# Check your answer here.
df[["retweet_count", "favorite_count", "is_retweet", "metrics"]]

retweet_count  favorite_count  is_retweet         metrics
0                72             462           0    (72, 462, 0)
1                50             361           0    (50, 361, 0)
2               135             870           0   (135, 870, 0)
3               374            1430           0  (374, 1430, 0)
4               137               0           1     (137, 0, 1)
...             ...             ...         ...             ...
3195           1197               0           1    (1197, 0, 1)
3196            622            2614           0  (622, 2614, 0)
3197            253            1942           0  (253, 1942, 0)
3198              2               6           0       (2, 6, 0)
3199            595               0           1     (595, 0, 1)

[3200 rows x 4 columns]

4\. Add a new column `num_mentions` to `df` such that each column value is the number of '@' symbols in the `text` column value. 

In [52]:
# Your answer here
df["num_mentions"] = df.text.apply(lambda x: x.count('@'))

In [53]:
# Check your answer here.
df[["text", "num_mentions"]]

text  num_mentions
0      It once held a lake. And what we're going to do, as soon as we get there, is to explore the rocks that were deposi… https://t.co/UW9wngvx4C             0
1      I asked the team [...] to come up with something that would symbolize, to mark this challenge and thank in particu… https://t.co/QH235r9zjq             0
2      Perseverance is the very first rover designed to seek signs of past microbial life, by collecting and caching rock… https://t.co/nn72OHRVv6             0
3     🔴 LIVE: Tune in for a preview of the next Mars landing! On Feb. 18, our @NASAPersevere rover will arrive at the Red… https://t.co/rHRhIrzXWR             1
4      RT @NASAMars: Don't forget to join us for today's briefing at 4:30pm ET (21:30 UTC): https://t.co/fGItMYWnFR Tag your questions #Countdown…             1
...                                                                                                                                            ...           ...
3195    RT @NASAKennedy: Looking for live views of the launch pad? We got you covered. 🚀👀Check out the rocket that will carry @AstroBehnken and @…             3
3196  For the first time since 2011, we’re sending American astronauts back to space, on an American rocket, from America… https://t.co/FsC1feO84c             0
3197    LIVE: Want to go behind-the-scenes of #LaunchAmerica?Join our experts on @reddit to ask questions about the missi… https://t.co/p7iqJY4jyX             1
3198                                                               @brandonleblanc @SpaceX @AmyShiraTeitel See you online! https://t.co/FaQ7K75sBP             3
3199  RT @NASAEarth: Ready to #LaunchAmerica? Tomorrow, two astronauts 👨🏻‍🚀👨🏻‍🚀 are launching 🚀 to the @Space_Station from @NASAKennedy’s histori…             3

[3200 rows x 2 columns]